[Refactor] Make `DisplayDriver` a true parent class; add `DisplayDriverFactory`

kdmukai opened the pull request:
## Description

This refactor does not create any user-facing changes. It is strictly a restructuring to enable future feature changes to display drivers.

Prior to this PR, `DisplayDriver` was already a factory class that would explicitly instantiate the requested display driver (ST7789, ST7789_mpy, ili9341). I'm not quite sure what I was thinking in that original implementation as it's somewhat obvious that I did mean for `DisplayDriver` to eventually be a parent base class (e.g. defining a `show_image()` method that then internally makes the same call to its `self.display` member var). I think the original implementation trying to be both a factory and a parent class at the same time threw me off.

So this PR rectifies that.

The part of the prior `DisplayDriver` class that handled actual driver instantiation has been moved into the new `DisplayDriverFactory`.

And then the rest of the prior `DisplayDriver` class that defined the driver interface is now made into a proper parent class (`BaseDisplayDriver`). `ST7789` (our original driver for the standard Waveshare hat) and `ST7789_mpy` (driver for the SeedSigner+ 320x240 hat) have been updated to be child classes of the new `DisplayDriver`.

I also made a cursory attempt at making the same changes to `ILI9341` but I haven't tested it as I don't have a hardware setup for it at the moment. Its inclusion in the display driver list was always a "eh, why not?" at this stage but was only intended to facilitate hardware R&D (it's specified as "beta" anyway).

### Motivation
I've been working on separate R&D efforts around user-selected gamma curves in the display drivers and controlling the screen backlight via pwm. In both cases, it was clear that having a common `BaseDisplayDriver` parent class was an improvement to the code structure.

But both of those efforts are in limbo, unsure of how to proceed.

So instead I pulled out just the factory/parent/child changes for this PR. Otherwise I was worried that this useful refactor would get lost as the gamma curve and pwm backlight branches languished.

_Note: I did not use python's built-in abstract base class (`abc.ABC`) because we haven't used it for any other abstract base classes (e.g. `BaseScreen`) and `ABC` is not supported in MicroPython. Using the `ABC` class isn't much of a "win" as far as I can tell, anyway._

---

This pull request is categorized as a:
- [x] Code refactor

## Checklist
- [x] I’ve run `pytest` and made sure all unit tests pass before submitting the PR

If you modified or added functionality/workflow, did you add new unit tests?
- [x] N/A

I have tested this PR on the following platforms/os:
- [x] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board

kdmukai committed:
Refactor `DisplayDriver` to be a true parent class; add Factory

kdmukai committed:
Rename to `BaseDisplayDriver`; ili9341 changes

alvroble commented:
I tested this with the SeedSigner OS. Previously, if I selected the ILI9341 display while using an ST7789 screen, the interface behaved strangely but didn’t crash, it just looked incorrect. Now it seems to crash and requires a device reset.

It’s an issue worth solving before merging. Users could accidentally select the wrong display type (for example, due to a finger slip or confusion), so it should ideally fail more gracefully.

alvroble commented:
I don’t have an ILI9341 display to test with, but it might be worth checking this branch with that display too, in case something in the initialization flow is now broken. I couldn’t identify any obvious issue in the code.

newtonick commented:
Like @alvroble commented, ILI9341 causes a crash. The issue appears to be because `self._rst` can't be found. I believe the root cause is because `ILI9341` is intended to be a `@dataclass` but is never declared as one. So `__post_init__` never executes.

Error:

```python
Traceback (most recent call last):
  File "/home/pi/seedsigner/src/seedsigner/controller.py", line 318, in start
    next_destination = next_destination.run()
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 168, in run
    return self._run_view()
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 163, in _run_view
    return self.view.run()
  File "/home/pi/seedsigner/src/seedsigner/views/settings_views.py", line 234, in run
    self.renderer.initialize_display()
  File "/home/pi/seedsigner/src/seedsigner/gui/renderer.py", line 51, in initialize_display
    self.disp = DisplayDriverFactory.instantiate_display_driver(self.display_type, width=int(width), height=int(height))
  File "/home/pi/seedsigner/src/seedsigner/hardware/displays/display_driver.py", line 87, in instantiate_display_driver
    display.begin()
  File "/home/pi/seedsigner/src/seedsigner/hardware/displays/ili9341.py", line 300, in begin
    self.reset()
  File "/home/pi/seedsigner/src/seedsigner/hardware/displays/ili9341.py", line 197, in reset
    if self._rst is not None:
AttributeError: 'ILI9341' object has no attribute '_rst'
2025-12-19 18:01:06,181    ERROR [seedsigner.controller handle_exception (446)]: 'ILI9341' object has no attribute '_rst'
Traceback (most recent call last):
  File "/home/pi/seedsigner/src/seedsigner/controller.py", line 318, in start
    next_destination = next_destination.run()
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 168, in run
    return self._run_view()
  File "/home/pi/seedsigner/src/seedsigner/views/view.py", line 163, in _run_view
    return self.view.run()
  File "/home/pi/seedsigner/src/seedsigner/views/settings_views.py", line 234, in run
    self.renderer.initialize_display()
  File "/home/pi/seedsigner/src/seedsigner/gui/renderer.py", line 51, in initialize_display
    self.disp = DisplayDriverFactory.instantiate_display_driver(self.display_type, width=int(width), height=int(height))
  File "/home/pi/seedsigner/src/seedsigner/hardware/displays/display_driver.py", line 87, in instantiate_display_driver
    display.begin()
  File "/home/pi/seedsigner/src/seedsigner/hardware/displays/ili9341.py", line 300, in begin
    self.reset()
  File "/home/pi/seedsigner/src/seedsigner/hardware/displays/ili9341.py", line 197, in reset
    if self._rst is not None:
AttributeError: 'ILI9341' object has no attribute '_rst'
2025-12-19 18:01:06,193     INFO [seedsigner.controller start (363)]: Appending next destination: UnhandledExceptionView({'error': ['AttributeError', 'ili9341.py, 197, in reset', " 'ILI9341' object has no attribute '_rst'"]}) | clear_history: True
2025-12-19 18:01:06,194     INFO [seedsigner.controller start (313)]:
back_stack: [
     0: UnhandledExceptionView({'error': ['AttributeError', 'ili9341.py, 197, in reset', " 'ILI9341' object has no attribute '_rst'"]}) | clear_history: True
]
2025-12-19 18:01:06,195     INFO [seedsigner.controller start (317)]: Executing UnhandledExceptionView({'error': ['AttributeError', 'ili9341.py, 197, in reset', " 'ILI9341' object has no attribute '_rst'"]}) | clear_history: True
2025-12-19 18:01:06,200  WARNING [seedsigner.gui.components __post_init__ (515)]: Supersampling disabled for large font size: 20
```

@kdmukai did you mean to define `ILI9341` as a dataclass in this refector? I believe all that is needed is to add the import (`from dataclasses import dataclass`) and `@dataclass` above `class ILI9341(BaseDisplayDriver):`.

kdmukai committed:
bufixes: dataclass; param

kdmukai commented:
Rebased to current`dev`.

Fixed missing `dataclass` decorator noted by @newtonick.

Fixed another bug found with a kwarg name mismatch in `def invert()`

newtonick commented:
ACK and tested.
